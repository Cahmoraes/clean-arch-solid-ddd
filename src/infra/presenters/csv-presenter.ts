import type { Presenter } from './presenter'

export class CSVPresenter implements Presenter<string> {
  public format<Input extends Array<any>>(data: Input): string {
    const headers = this.extractHeaders(data)
    const rows = this.extractValues(data)
    return `${headers}\n${rows}`
  }

  private extractHeaders(headers: any): string {
    return Object.keys(headers[0]).join(',')
  }

  private extractValues(rows: any[]): string {
    const values: any[] = []
    for (const [, rowValues] of rows.entries()) {
      for (const value of Object.values(rowValues)) {
        console.log(value)
        values.push(value)
      }
    }
    return values.join(',')
  }
}
