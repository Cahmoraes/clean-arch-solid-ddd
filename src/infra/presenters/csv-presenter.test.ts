import { CSVPresenter } from './csv-presenter'

describe('CSV Presenter', () => {
  let sut: CSVPresenter

  const data = [
    {
      id: 'fake_id',
      role: 'MEMBER',
      createdAt: '2025-01-18T23:48:24.009Z',
      name: 'any_name',
      email: 'any_email',
    },
    {
      id: '53524955-138e-43bc-8939-0c80e32c79ed',
      role: 'MEMBER',
      createdAt: '2025-01-19T10:31:45.365Z',
      name: 'Ms. Madeline Yost',
      email: 'Shanelle_McLaughlin@gmail.com',
    },
  ]

  beforeEach(() => {
    sut = new CSVPresenter()
  })

  test('Deve formatar para CSV', () => {
    const result = sut.format(data)
    expect(result).toMatchSnapshot()
  })
})
