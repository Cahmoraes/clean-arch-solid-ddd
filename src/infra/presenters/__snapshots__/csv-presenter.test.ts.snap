// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CSV Presenter > Deve formatar para CSV 1`] = `
"id,role,createdAt,name,email
fake_id,MEMBER,2025-01-18T23:48:24.009Z,any_name,any_email,53524955-138e-43bc-8939-0c80e32c79ed,MEMBER,2025-01-19T10:31:45.365Z,Ms. Madeline Yost,Shanelle_McLaughlin@gmail.com"
`;
